### PR TITLE
Refactor HMC

### PR DESCRIPTION
**Pull request by ColCarroll**
Started refactoring Hamiltonian methods into a common base class.  This lets `HamiltonianMC` share some theano machinery from NUTS, and provides a ~2x speedup using the benchmark from #1570.  The same PR also shows that I haven't changed either algorithm at all.

My larger plan is to use the approach from @betanalpha's exhaustive monte carlo paper (#1497) to provide different theano functions for expanding trajectory, checking trajectory, and checking termination so that HMC, NUTS, and XHMC can share (optimized) machinery, and perhaps there can be easy experiments with further mixing of methods.  I'm hoping to have a PR soon where `BaseHMC` implements algorithm 2 from the paper as `astep`, so that the implementation of `NUTS` and `HamiltonianMC` is only an `__init__` which prepares the appropriate functions.

Also, note that this will be merged into 3.1, not 3.0.

### CODE REVIEW

**Comment by ColCarroll** [comment]
Ah frig, I think we need to rebase 3.1 off master, then I will rebase this.  Happy for thoughts in the meantime.

File: pymc3/step_methods/hmc/base_hmc.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+class BaseHMC(ArrayStepShared):
+    default_blocked = True
+
+    def __init__(self, vars=None, scaling=None, step_scale=0.25, is_cov=False,
+                 model=None, blocked=True, use_single_leapfrog=False, **theano_kwargs):
+        """
+        Parameters
+        ----------
+            vars : list of theano variables
```
Comment (twiecki) [inline pymc3/step_methods/hmc/base_hmc.py]: Don't need to indent that level.

**Comment by twiecki** [comment]
I skimmed over it, looks like a good refactor. Seems like you didn't touch much of the math, correct?

**Comment by ColCarroll** [comment]
No math changes -- I brought a test case in from master to make sure the sampling stayed exactly the same.  The only change an end user will see from this is the performance speedup in `HamiltonianMC`.

**Comment by twiecki** [comment]
Sorry, I rebased the 3.1 branch off of master and resolved a conflict, looks like you'll have to rebase this one.

**Comment by twiecki** [comment]
I attempted to do the merge myself, can you see if I messed anything up: https://github.com/pymc-devs/pymc3/commit/ea82ebd554c426d4af85881a39e1698805cf47c8 https://github.com/pymc-devs/pymc3/commit/140a80c8f2a6949c90ba935d0f1199afa46b1fc0

**Pull request closed without merging**